cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORParser.java

```diff
@@ -615,17 +732,25 @@ public JsonToken nextToken() throws IOException
         // as well as handle names for Object entries.
         if (_streamReadContext.inObject()) {
             if (_currToken != JsonToken.FIELD_NAME) {
-                _tagValue = -1;
+                _tagValues.clear();
                 // completed the whole Object?
                 if (!_streamReadContext.expectMoreValues()) {
+                    --_nestedDepth;
+                    if (!_stringRefs.empty() && _stringRefs.peek().depth == _nestedDepth) {
+                        _stringRefs.pop();
+                    }
                     _streamReadContext = _streamReadContext.getParent();
                     return (_currToken = JsonToken.END_OBJECT);
                 }
                 return (_currToken = _decodePropertyName());
             }
         } else {
             if (!_streamReadContext.expectMoreValues()) {
-                _tagValue = -1;
+                --_nestedDepth;
+                if (!_stringRefs.empty() && _stringRefs.peek().depth == _nestedDepth) {
```

Comment:
Using a simple sub-class of `Stack` would allow hiding this logic into single method.

Reply:
I created a helper class for this. I chose to use containment rather than subclass for this case as the interface is slightly different.

Reply:
@here-abarany Ok yes, that is cleaner approach from OO design perspective. Although being only used internally not a big thing either way.